### PR TITLE
Rework output handling

### PR DESCRIPTION
**Pull request by dertseha**
## Why

This PR is a proposal to better separate stdout/stderr usage.
Primary motivation for this was the recommendation from [clig.dev](https://clig.dev/#the-basics) to send messages to `stderr` and have `stdout` only for net information from the tasks/tools.
Since the change is an incompatible one, further, related incompatibilities are included, see below.

## What

1. Expand `Taskflow.Output` as a separate type having two writer, one for `Primary` output, one for `Message` output. (incompatible API change)
1. Goyek code was modified to only use `Message` output. (incompatible user-interface change) As a side-effect, any `fmt.Fprintf` error "handling" has been centralized and explicitly skipped.
1. Removed the `DefaultOutput` global variable. It seems as if this was only used for the tests. Tests should override this explicitly anyway, to avoid side-effects. Furthermore, such global defaults are tricky; I refer to the potential problems as seen with `http.Client` and the transport defaults. (incompatible API change)

Documentation/Changelog has not yet been updated, would do so if proposed changes are discussed and settled.

## Checklist

- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated.
- [x] The code changes follow [Effective Go](https://golang.org/doc/effective_go).

### CODE REVIEW

**Comment by pellared** [comment]
I just quickly scrolled the PR. First thoughts:
1. What do you think of adding a new field e.g. ErrWriter instead of changing the `Output` field.
2. I think that the "usage" should be printed to stdout by default. Reasoning: https://stackoverflow.com/questions/2199624/should-the-command-line-usage-be-printed-on-stdout-or-stderr
3. Do you have a use case or example where it could be handy? Can you check how `go test` works? Is it not printing everything to `stdout`?

**Comment by dertseha** [comment]
Thank you for your initial response.

1. Thoughts against a separate field:
The two writers are used together and passed on together. So I saw them as a data couple deserving a separate type.
Further, I wanted to explictily communicate the clig.dev - inspired separation of "primary" output and "message" output.

2. Thank you for the hint on usage output, I've modified the code to write usage to primary output in case help was explicitly requested

3. This separation can be handy when wanting to process the output further. An example is the output of `golangci-lint`, when it is to be processed by another tool (such as an IDE). In this case I'd like to avoid any status messages from goyek.
This particular example is not the best, yet gives hint at why the separation is necessary.

**Comment by pellared** [comment]
@dertseha 

In order to make `goyek` IDE or navigation friendly, we should first add some kind of indication from where the log message was emitted. I mean something similar to what `go test` does (and linters as well):
```
=== RUN   Test_Register_errors
    taskflow_test.go:23: test
```

I see that `go` toolchain emits everything to `stdout`. We would need to reconsider if `goyek` should **consistently** follow `go` toolchain CLI "conventions" or https://clig.dev conventions. I know that you prefer https://clig.dev . I do not have a strong opinion but at the same time I do not feel convineced (yet) that following https://clig.dev  would bring benefits that are worth the little additional complexity. If would agree on using `stderr` and `stdout` then probably I would say that we should also follow the short+long flags as recommended in https://clig.dev

I need also to check how `stderr` behaves on PowerShell (I am not sure if it does not have some different formatting)...

PS. I do not say "no" I

**Comment by pellared** [comment]
Changing the `goyek.ps1` to the code below fixed the issue:

```
Push-Location "$PSScriptRoot\build" -ErrorAction "Stop"
& go run . -wd=".." $args
Pop-Location
exit $global:LASTEXITCODE
```

**Comment by pellared** [comment]
From  https://golang.org/cmd/go/

> All test output and summary lines are printed to the go command's standard output, even if the test printed them to its own standard error. (The go command's standard error is reserved for printing errors building the tests.

**Comment by pellared** [comment]
Regarding

> Thoughts against a separate field:
> The two writers are used together and passed on together. So I saw them as a data couple deserving a separate type.
> Further, I wanted to explictily communicate the clig.dev - inspired separation of "primary" output and "message" output.

Maybe we can even unexport it at this point? Personally, I had no use case for changing the output...

@dertseha We can make a dedicated release where we want to "adopt" https://clig.dev. What do you think about it?

**Comment by dertseha** [comment]
Calling the writer for `stdout` `Output` would end up in `Output().Output`. I could be convinced about `Output().Standard` or `Output().Main`.
As for describing and naming: Isn't the goal of naming to find one that matches the description?

I know this PR is running stale, I am hoping to be able to pick it up again soon; And need to anyway due to other PRs meanwhile.

**Comment by pellared** [comment]
What do you think of renaming `Output` to `Stdout` and adding `Stderr` field for messaging?

```go
flow.Stdout = myPrimaryWriter
flow.Stderr = myMessagingWriter
```

I think it might be more obvious and would match https://golang.org/pkg/os/exec/#Cmd.Stdout API

**Comment by dertseha** [comment]
I've brought the PR up to date to current `main`. I also renamed primary/message to standard/messaging.

I wanted to have the two writer be in their own type as they belong together and not move around two separate fields.

File: output.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	Standard io.Writer
+	// Messaging output is for any status information, such as logging or error messages.
+	Messaging io.Writer
+}
+
+// WriteMessagef prints the given format message and its arguments to the Messaging writer.
+// The result of this action is ignored.
+func (out Output) WriteMessagef(format string, a ...interface{}) {
+	writeLinef(out.Messaging, format, a...)
+}
```
Comment (pellared) [inline output.go:25]: 1. I am not sure that it is good to export this method.
2. Also it is not consistent that there is no parallel method for writing the primary/standard output.

File: tf.go [hunk 2]
```diff
@@ -17,7 +16,7 @@ import (
 type TF struct {
 	ctx         context.Context
 	name        string
-	writer      io.Writer
+	output      Output
```
Comment (pellared) [inline tf.go:19]: > // Standard output is for information that is expected from the executed tasks.

Shouldn't we just to pass the Standard writer here? Then we would use the Messaging just in `flowRunner` when bad usage is reported. The Messaging writer could still be handy in future e.g. if we allow using custom a flow runner.

**Comment by pellared** [comment]
@dertseha What is the latest on this PR? Do you need this functionality? Do you need some help with the implementation?

**Comment by dertseha** [comment]
@pellared Thank you for another ping, and your patience.
My original intent is/was to provide a consistent way to route through both output streams (for default output and messaging - also called stdout and stderr). As earlier identified, my attempt was to follow clig.dev , yet this doesn't seem to fully apply here, both from concept, and from vision/goal of the project. 

Added to that is my being-away state for several weeks and I no longer feel in touch with the project.

So I am open to close the PR and the topic can be picked up in a future idea.

**Pull request closed without merging**